src/ripple/protocol/impl/AccountID.cpp

```diff
@@ -17,17 +17,101 @@
 */
 //==============================================================================
 
+#include <ripple/basics/hardened_hash.h>
 #include <ripple/protocol/AccountID.h>
 #include <ripple/protocol/PublicKey.h>
 #include <ripple/protocol/digest.h>
 #include <ripple/protocol/tokens.h>
+#include <array>
 #include <cstring>
+#include <mutex>
 
 namespace ripple {
 
+namespace detail {
+
+/** Caches the base58 representations of AccountIDs */
+class AccountIdCache
+{
+private:
+    // The total number of items the cache buffer can contain
+    std::size_t const count_;
+
+    // The maximum size (in bytes) of an item, along with a null terminator
+    // and aligned at 8 bytes
+    std::size_t const size_ = 64;
+
+    // The offset of the string representation of an AccountID
+    std::size_t const offset_ = 24;
+
+    // We use a hash function designed to resist algorithmic complexity attacks
+    hardened_hash<> hasher_;
+
+    // The memory block
+    std::unique_ptr<char[]> chars_;
+
+    std::array<std::mutex, 16> mutex_;
+
+public:
+    AccountIdCache(std::size_t count)
+        : count_(count), chars_(std::make_unique<char[]>(count_ * size_))
+    {
+        std::memset(chars_.get(), 0, count_ * size_);
+    }
+
+    std::string
+    toBase58(AccountID const& id)
+    {
+        assert(chars_.get() != nullptr);
+
+        auto const index = hasher_(id.data()) % count_;
```

Comment:
Do we even need to hash the id? Why not use the low order uint64_t?

Reply:
We don't _really_ need to hash the whole thing, no. The cost to do it is likely not that big, but there are some things to watch out for: first, `hasher_` is a hardened hash (i.e. it resists algorithmic complexity attacks, by using tagged hashing) and we need that functionality.

Something like this might be sufficient, assuming `cookie_` is the `std::uint64_t` random tag:

    auto const index = (*reinterpret_cast<std::uint64_t const*>(id.data()) ^ cookie_) % count_;

Reply:
Yes, I think this would be sufficient for sure. The hardened hash is useful when attackers craft ID to cause collisions in hash maps and slow lookups because of excessive probing. Here we don't really use a hash map with probing, but overwrite the first entry if it doesn't match what we look for, so the worse case scenario would be that the cache is not effective, which I don't think would be crippling.
So I think the xor with a randomly generated value should be plenty safe!